File: native-apps/macos/NativeSettings.swift
import Foundation

struct NativeSettings: Codable {
    /// macOS-ийн domain шугам. 2026-08-12-нд асав: DNS (wildcard), nginx vhost,
    /// TLS гэрчилгээ, API-гийн origin allowlist дөрвүүлэн бэлэн.
    ///
    /// Web ба API нэг host дээр байгаа нь санаатай: ажлын мужаас гарах дуудлага
    /// same-origin болж, session cookie нь `SameSite=Strict` хэвээр ажиллана.
    /// Бүртгэл: `native-apps/shared/device_lines.json`.
    static let lineOrigin = "https://mac.nexus.gerege.mn"

    var schemaVersion = 4
    var launchAtLogin = false
    var language = "mn"
    var webEndpoint = NativeSettings.lineOrigin
    var apiEndpoint = NativeSettings.lineOrigin
    var printerTransport = "USB"
    var printerHost = ""
    var printerPort = 9100
    var serialPort = ""
    var baudRate = 9600
    var paperWidth = "80 mm"
    var scannerMode = "Keyboard wedge"
    var scannerSuffix = "Enter"
    var biometricLock = true
    var idleLockMinutes = 5
    var updateChannel = "Stable"
    var telemetry = true
    var deviceName = Host.current().localizedName ?? "Mac"
    var site = ""
    var deviceID = ""

    static let storageKey = "mn.gerege.nexus.native-settings.v1"
    static func load() -> NativeSettings {
        guard let data = UserDefaults.standard.data(forKey: storageKey),
              var value = try? JSONDecoder().decode(Self.self, from: data) else { return Self() }
        if value.schemaVersion < 2 {
            if value.webEndpoint == "http://localhost:3000" { value.webEndpoint = "https://nexus.gerege.mn" }
            if value.apiEndpoint == "http://localhost:8080" { value.apiEndpoint = "https://nexus.gerege.mn" }
            value.schemaVersion = 2
            value.save()
        }
        // v4: schemaVersion-ыг зөвхөн урагшлуулна. Хадгалагдсан хаягийг
        // хөндөхгүй — шугам асахаас өмнөх богино хугацаанд хадгалагдсан утга
        // (`https://nexus.gerege.mn`) нь ижил backend руу очдог тул ажилласаар
        // байна. Түүнийг хүчээр зөөвөл хөтчийн шугамыг САНААТАЙ сонгосон
        // суулгацыг булааж авах болно; шинэ суулгац `lineOrigin`-оор эхэлнэ.
        if value.schemaVersion < 4 {
            value.schemaVersion = 4
            value.save()
        }
        return value
    }
    func save() {
        guard let data = try? JSONEncoder().encode(self) else { return }
        UserDefaults.standard.set(data, forKey: Self.storageKey)
    }
}

File: native-apps/desktop/windows/NativeSettings.cs
using System;
using System.IO;
using System.Text.Json;

namespace GeregeNexusNativeWin;

public sealed class NativeSettings
{
    public int SchemaVersion { get; set; } = 3;
    public bool LaunchAtLogin { get; set; }
    public string Language { get; set; } = "mn";
    // Web ба API нэг host дээр байгаа нь санаатай: ажлын мужаас гарах дуудлага
    // same-origin болж, session cookie нь SameSite=Strict хэвээр ажиллана.
    public string WebEndpoint { get; set; } = ShellProfile.LineOrigin;
    public string ApiEndpoint { get; set; } = ShellProfile.LineOrigin;
    public string PrinterTransport { get; set; } = "USB";
    public string PrinterHost { get; set; } = "";
    public int PrinterPort { get; set; } = 9100;
    public string PaperWidth { get; set; } = "80 mm";
    public string ScannerMode { get; set; } = "Keyboard wedge";
    public string ScannerSuffix { get; set; } = "Enter";
    public string SerialPort { get; set; } = "";
    public int BaudRate { get; set; } = 9600;
    public bool BiometricLock { get; set; } = true;
    public int IdleLockMinutes { get; set; } = 5;
    public string UpdateChannel { get; set; } = "Stable";
    public bool Telemetry { get; set; } = true;
    public string DeviceName { get; set; } = Environment.MachineName;
    public string Site { get; set; } = "";
    public string DeviceId { get; set; } = "";
    public int DrawerPulseMs { get; set; } = 120;

    private static string FilePath => Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "GeregeNexus", "native-settings-v1.json");
    public static NativeSettings Load()
    {
        NativeSettings value;
        try { value = JsonSerializer.Deserialize<NativeSettings>(File.ReadAllText(FilePath)) ?? new(); }
        catch { return new(); }

        // v3: dev-ийн localhost анхдагчийг энэ build-ийн domain шугам руу
        // зөөнө. `https://nexus.gerege.mn` энд зориудаар байхгүй — тэр хаяг
        // ижил backend руу очдог тул ажилласаар байгаа бөгөөд түүнийг хүчээр
        // зөөвөл хөтчийн шугамыг санаатай сонгосон суулгацыг булааж авна.
        if (value.SchemaVersion < 3)
        {
            if (value.WebEndpoint is "http://localhost:3000") value.WebEndpoint = ShellProfile.LineOrigin;
            if (value.ApiEndpoint is "http://localhost:8080") value.ApiEndpoint = ShellProfile.LineOrigin;
            value.SchemaVersion = 3;
            value.Save();
        }
        return value;
    }

    public void Save()
    {
        Directory.CreateDirectory(Path.GetDirectoryName(FilePath)!);
        File.WriteAllText(FilePath, JsonSerializer.Serialize(this, new JsonSerializerOptions { WriteIndented = true }));
    }
}
